docs/hugo/content/design/ADR-2022-09-Multiple-Credential-Operator.md

```diff
@@ -0,0 +1,153 @@
+
+NOTE: If namespaced secret or global secret is used, operator would inject `serviceoperator.azure.com/credential-from` annotation with the secret name value. 
+
+**Fixed namespaced secret name pattern:**
+```
+Namespaced secret -> aso-credential (in the namespace)
+```
+
+**Annotation for per-resource-group and per-resource secret**
+```
+annotations: 
+  serviceoperator.azure.com/credential-from: any-secret
+```
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly if using a namespaced secret
+2. Users get the flexibility of over-riding the credential they want to use as per resource and as per resource group
+
+**Cons:**
+1. Could be complicated/confusing for users to understand the usage and credentials being used
+
+### Option 4: Secret lookup + configuration using global secret
+
+This option, relates to the option 3. However, the change is how we configure to use the per-resource-group and per-resource secret. We can have a unique slice property(CREDENTIAL_FROM) in the global secret where admin users can specify the secrets they want to load on the startup.
+Then only these loaded secrets would be used further for per-resource and per-resource-group. This option would restrict the other non-admin users(who don't have access to the global secret) to add their credential. 
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly if using a namespaced secret
+2. Security conscious, not anyone in the cluster can add to use their credentials.
+
+**Cons:**
+1. Could be complicated/confusing for users to understand the usage and credentials being used
+2. Not very flexible, as pod will have to startup again if a new secret needs to be loaded
+
+## Handling for the failure cases and logging
+
+- Failure on loading a secret would result in failure on resource creation
+- Failure on authenticating the credentials would result in failure on resource creation
+- Logging should be clear about which credential is being used for a resource
+- Logging should be clear about the credential load/authentication failure
+
+## Milestones
+
+### Support for namespaced secrets
+
+We aim to do the initial work needed in the operator to support this type of multitenancy and add support for namespaced secrets.
+
+**Timelines:** TBD
+
+### Support for per-resource-group secrets
+
+We aim to support per-resource-group secrets under this milestone and configuration work required.
+
+**Timelines:** TBD
+
+### Support for per-resource secrets
+
+We aim to support per-resource secrets under this milestone and configuration work required.
+
+**Timelines:** TBD
+
+### Support for credential cache
+
+TBD
+
+## Open questions
+1. Do we need to support both per-resource and per-resource-group credentials?
+2. Do we want a serviceoperator.azure.com/subscription-id annotation that we write? We can use this to protect against users accidentally updating credential to move resources between subscriptions.
```

Comment:
IMO yes (see my earlier comments). I think this is an important protection. It helps even the current global credential case. User can always get around it if they want by manually clearing that annotation.